File: frontend/src/services/projectUserService.js
import apiClient, {
  API_PROJECT_USERS,
  API_PROJECT_USERS_COUNT,
  API_PROJECT_USERS_BY_PROJECT,
  API_PROJECT_USERS_BY_USER,
  API_PROJECT_USERS_COUNT_BY_PROJECT,
  API_PROJECT_USERS_COUNT_BY_USER,
  API_PROJECT_USER_BY_ID,
} from './api';

class ProjectUserService {
  // Create a new project-user relationship
  async createProjectUser(projectUserData) {
    try {
      const response = await apiClient.post(API_PROJECT_USERS, projectUserData);
      return response.data;
    } catch (error) {
      console.error('Error creating project-user relationship:', error);
      throw error;
    }
  }

  // Get all project-user relationships with pagination
  async getProjectUsers(skip = 0, limit = 10) {
    try {
      const response = await apiClient.get(API_PROJECT_USERS, {
        params: { skip, limit }
      });
      return response.data;
    } catch (error) {
      console.error('Error fetching project-user relationships:', error);
      throw error;
    }
  }

  // Get project-user relationship by ID
  async getProjectUserById(projectUserId) {
    try {
      const response = await apiClient.get(API_PROJECT_USER_BY_ID(projectUserId));
      return response.data;
    } catch (error) {
      console.error('Error fetching project-user relationship:', error);
      throw error;
    }
  }

  // Update project-user relationship
  async updateProjectUser(projectUserId, projectUserData) {
    try {
      const response = await apiClient.put(API_PROJECT_USER_BY_ID(projectUserId), projectUserData);
      return response.data;
    } catch (error) {
      console.error('Error updating project-user relationship:', error);
      throw error;
    }
  }

  // Delete project-user relationship
  async deleteProjectUser(projectUserId) {
    try {
      const response = await apiClient.delete(API_PROJECT_USER_BY_ID(projectUserId));
      return response.data;
    } catch (error) {
      console.error('Error deleting project-user relationship:', error);
      throw error;
    }
  }

  // Get project-user relationships count
  async getProjectUsersCount() {
    try {
      const response = await apiClient.get(API_PROJECT_USERS_COUNT);
      return response.data;
    } catch (error) {
      console.error('Error fetching project-user relationships count:', error);
      throw error;
    }
  }

  // Get project-user relationships by project
  async getProjectUsersByProject(projectId, skip = 0, limit = 10) {
    try {
      const response = await apiClient.get(API_PROJECT_USERS_BY_PROJECT(projectId), {
        params: { skip, limit }
      });
      return response.data;
    } catch (error) {
      console.error('Error fetching project-user relationships by project:', error);
      throw error;
    }
  }

  // Get project-user relationships by user
  async getProjectUsersByUser(userId, skip = 0, limit = 10) {
    try {
      const response = await apiClient.get(API_PROJECT_USERS_BY_USER(userId), {
        params: { skip, limit }
      });
      return response.data;
    } catch (error) {
      console.error('Error fetching project-user relationships by user:', error);
      throw error;
    }
  }

  // Get project-user relationships count by project
  async getProjectUsersCountByProject(projectId) {
    try {
      const response = await apiClient.get(API_PROJECT_USERS_COUNT_BY_PROJECT(projectId));
      return response.data;
    } catch (error) {
      console.error('Error fetching project-user relationships count by project:', error);
      throw error;
    }
  }

  // Get project-user relationships count by user
  async getProjectUsersCountByUser(userId) {
    try {
      const response = await apiClient.get(API_PROJECT_USERS_COUNT_BY_USER(userId));
      return response.data;
    } catch (error) {
      console.error('Error fetching project-user relationships count by user:', error);
      throw error;
    }
  }
}

export default new ProjectUserService(); 

File: backend/src/services/project_user_service.py
from sqlalchemy.orm import Session
from sqlalchemy.exc import IntegrityError
from src.model.project_user import ProjectUser, ProjectUserCreate, ProjectUserResponse
from typing import List, Optional
import logging

logger = logging.getLogger(__name__)

class ProjectUserService:
    """Service class for handling project-user relationship business logic."""
    
    def __init__(self, db: Session):
        self.db = db
    
    def create_project_user(self, project_user_data: ProjectUserCreate) -> ProjectUserResponse:
        try:
            db_project_user = ProjectUser(**project_user_data.model_dump())
            self.db.add(db_project_user)
            self.db.commit()
            self.db.refresh(db_project_user)
            logger.info(f"Created project-user relationship with ID: {db_project_user.id}")
            return ProjectUserResponse.model_validate(db_project_user)
        except IntegrityError as e:
            self.db.rollback()
            logger.error(f"Failed to create project-user relationship due to integrity error: {e}")
            raise ValueError("Project-user relationship creation failed: integrity constraint violation")
        except Exception as e:
            self.db.rollback()
            logger.error(f"Failed to create project-user relationship: {e}")
            raise ValueError(f"Project-user relationship creation failed: {str(e)}")
    
    def get_project_user_by_id(self, project_user_id: int) -> Optional[ProjectUserResponse]:
        project_user = self.db.query(ProjectUser).filter(ProjectUser.id == project_user_id).first()
        if project_user:
            return ProjectUserResponse.model_validate(project_user)
        return None
    
    def get_project_users_by_project(self, project_id: int, skip: int = 0, limit: int = 10) -> List[ProjectUserResponse]:
        project_users = self.db.query(ProjectUser).filter(
            ProjectUser.project_id == project_id
        ).offset(skip).limit(limit).all()
        logger.info(f"Retrieved {len(project_users)} project-user relationships for project {project_id}")
        return [ProjectUserResponse.model_validate(pu) for pu in project_users]
    
    def get_project_users_by_user(self, user_id: int, skip: int = 0, limit: int = 10) -> List[ProjectUserResponse]:
        project_users = self.db.query(ProjectUser).filter(
            ProjectUser.user_id == user_id
        ).offset(skip).limit(limit).all()
        logger.info(f"Retrieved {len(project_users)} project-user relationships for user {user_id}")
        return [ProjectUserResponse.model_validate(pu) for pu in project_users]
    
    def list_project_users(self, skip: int = 0, limit: int = 10) -> List[ProjectUserResponse]:
        project_users = self.db.query(ProjectUser).offset(skip).limit(limit).all()
        logger.info(f"Retrieved {len(project_users)} project-user relationships with skip={skip}, limit={limit}")
        return [ProjectUserResponse.model_validate(pu) for pu in project_users]
    
    def update_project_user(self, project_user_id: int, project_user_data: ProjectUserCreate) -> Optional[ProjectUserResponse]:
        db_project_user = self.db.query(ProjectUser).filter(ProjectUser.id == project_user_id).first()
        if not db_project_user:
            return None
        try:
            for key, value in project_user_data.model_dump().items():
                setattr(db_project_user, key, value)
            self.db.commit()
            self.db.refresh(db_project_user)
            logger.info(f"Updated project-user relationship with ID: {db_project_user.id}")
            return ProjectUserResponse.model_validate(db_project_user)
        except IntegrityError as e:
            self.db.rollback()
            logger.error(f"Failed to update project-user relationship {project_user_id} due to integrity error: {e}")
            raise ValueError("Update failed: integrity constraint violation")
        except Exception as e:
            self.db.rollback()
            logger.error(f"Failed to update project-user relationship {project_user_id}: {e}")
            raise ValueError(f"Update failed: {str(e)}")
    
    def delete_project_user(self, project_user_id: int) -> bool:
        db_project_user = self.db.query(ProjectUser).filter(ProjectUser.id == project_user_id).first()
        if not db_project_user:
            return False
        try:
            self.db.delete(db_project_user)
            self.db.commit()
            logger.info(f"Deleted project-user relationship with ID: {project_user_id}")
            return True
        except Exception as e:
            self.db.rollback()
            logger.error(f"Failed to delete project-user relationship {project_user_id}: {e}")
            raise ValueError(f"Deletion failed: {str(e)}")
    
    def get_project_users_count(self) -> int:
        return self.db.query(ProjectUser).count()
    
    def get_project_users_count_by_project(self, project_id: int) -> int:
        return self.db.query(ProjectUser).filter(ProjectUser.project_id == project_id).count()
    
    def get_project_users_count_by_user(self, user_id: int) -> int:
        return self.db.query(ProjectUser).filter(ProjectUser.user_id == user_id).count() 